sklearn/svm/_classes.py

```diff
@@ -1,14 +1,35 @@
 from numbers import Integral, Real
+import warnings
 
 import numpy as np
 
-from ._base import _fit_liblinear, BaseSVC, BaseLibSVM
+from ._base import _fit_liblinear, _get_liblinear_solver_type, BaseSVC, BaseLibSVM
 from ..base import BaseEstimator, RegressorMixin, OutlierMixin
 from ..linear_model._base import LinearClassifierMixin, SparseCoefMixin, LinearModel
 from ..utils import deprecated
 from ..utils.validation import _num_samples
 from ..utils.multiclass import check_classification_targets
-from ..utils._param_validation import Interval, StrOptions
+from ..utils._param_validation import Interval, StrOptions, Hidden
+
+
+def _choose_dual_automatically(loss, penalty, multi_class, X):
+    """Choose dual parameter value wrt to loss, penalty and shape of the data"""
+    if X.shape[0] < X.shape[1]:
+        try:
+            # check if the combination of parameters supports
+            # dual formulation
+            _get_liblinear_solver_type(multi_class, penalty, loss, True)
+            return True
+        except ValueError:
+            return False
+    else:
+        try:
+            # check if the combination of parameters doesn't support
+            # dual formulation
+            _get_liblinear_solver_type(multi_class, penalty, loss, False)
+            return False
+        except ValueError:
+            return True
```

Comment:
I'm a bit surprised here. I would have thought that the primal formulation is supported by all combinations. Do you know which one is not supported?

Reply:
I think "hinge" + "l2" and "epsilon_insensitive" + "l2"
https://github.com/glevv/scikit-learn/blob/7021bde4616b9a93c4f0a69d057774fc6a6f8115/sklearn/svm/_base.py#L1026-L1033
See [b2aa088](https://github.com/scikit-learn/scikit-learn/commit/b2aa0885f6a840dbb1b35759973db788e4d34d15)

Reply:
Thanks good to know.